utils/cas_version_gen

```diff
@@ -0,0 +1,105 @@
+#!/bin/bash
+#
+# Copyright(c) 2020 Intel Corporation
+# SPDX-License-Identifier: BSD-3-Clause-Clear
+#
+
+
+THIS=$(basename "$0")
+SOURCES_DIR="$(realpath ../)"
+MANUAL_VERSION_INPUT="$SOURCES_DIR/version"
+VERSION_FILE="$SOURCES_DIR/.metadata/cas_version"
+SUBMODULES=(
+    "ocf"
+    "test/functional/test-framework"
+)
+
+
+warning() {
+    echo -e "\e[33mWARNING\e[0m: $THIS: $*" >&2
+}
+
+error() {
+    echo -e "\e[31mERROR\e[0m: $THIS: $*" >&2
+    exit 1
+}
+
+
+# Check if we're inside a git repository
+if [[ -d "$SOURCES_DIR/.git" ]] && which git &>/dev/null &&\
+        (cd "$SOURCES_DIR" && git rev-parse --is-inside-work-tree &>/dev/null); then
+    if [[ ! -r "$MANUAL_VERSION_INPUT" ]]; then
+        error "can't read version input file"
+    fi
+    . "$MANUAL_VERSION_INPUT"
+    if [[ ! "$CAS_VERSION_MAIN" || ! "$CAS_VERSION_MAJOR" || ! "$CAS_VERSION_MINOR" ]]; then
+        error "wrong version input file format"
+    fi
+
+    CAS_VERSION_BUILD=$(cd "$SOURCES_DIR" && git log --merges --oneline | wc -l)
+    LAST_COMMIT_HASH=$(cd "$SOURCES_DIR" && git log -1 --pretty=format:%H)
+    LAST_COMMIT_HASH_ABBR=$(cd "$SOURCES_DIR" && git log -1 --pretty=format:%h)
+    LAST_COMMIT_DATE=$(cd "$SOURCES_DIR" && git log -1 --pretty=format:%ci |\
+                       sed "s/ /T/" | sed "s/ //" | sed "s/00$/:00/")
+    LAST_COMMIT_TIMESTAMP=$(cd "$SOURCES_DIR" && git log -1 --pretty=format:%ct)
+    for SUBMOD in ${SUBMODULES[@]}; do
+        LAST_SUB_COMMIT_HASHES+=($(cd "$SOURCES_DIR/$SUBMOD" && git log -1 --pretty=format:%H))
+        LAST_SUB_COMMIT_HASHES_ABBR+=($(cd "$SOURCES_DIR/$SUBMOD" && git log -1 --pretty=format:%h))
+    done
+    if [[ $(cd "$SOURCES_DIR" && git tag --points-at HEAD) ]]; then
+        CAS_VERSION_RELEASE="release"
+    elif [[ $(cd "$SOURCES_DIR" && git log -1 --pretty=format:%H)\
+            == $(cd "$SOURCES_DIR" && git log -1 --merges --pretty=format:%H) ]]; then
+        CAS_VERSION_RELEASE="master"
+    else
+        CAS_VERSION_RELEASE="devel"
+    fi
+
+    CAS_VERSION=$(printf "%02d.%02d.%01d.%04d.%s" $CAS_VERSION_MAIN $CAS_VERSION_MAJOR\
+                $CAS_VERSION_MINOR $CAS_VERSION_BUILD $CAS_VERSION_RELEASE)
+
+    mkdir -p $(dirname "$VERSION_FILE")
+    if ! touch "$VERSION_FILE"; then
+        error "couldn't create version file"
+    fi
+    echo "CAS_VERSION_MAIN=$CAS_VERSION_MAIN" > "$VERSION_FILE"
+    echo "CAS_VERSION_MAJOR=$CAS_VERSION_MAJOR" >> "$VERSION_FILE"
+    echo "CAS_VERSION_MINOR=$CAS_VERSION_MINOR" >> "$VERSION_FILE"
+    echo "CAS_VERSION_BUILD=$CAS_VERSION_BUILD" >> "$VERSION_FILE"
+    echo "CAS_VERSION_RELEASE=$CAS_VERSION_RELEASE" >> "$VERSION_FILE"
+    echo "CAS_VERSION=$CAS_VERSION" >> "$VERSION_FILE"
+    echo "LAST_COMMIT_HASH=$LAST_COMMIT_HASH" >> "$VERSION_FILE"
+    echo "LAST_COMMIT_HASH_ABBR=$LAST_COMMIT_HASH_ABBR" >> "$VERSION_FILE"
```

Comment:
Do we need both abbreviated and full hash? I think abbreviation is constructed in such way that it is unique across repo.

Reply:
I've added full hash for precision and abbreviated for convenience. Let me know if I should remove one of them.

Reply:
And also I don't know if the abbreviation is globally unique.

Reply:
Well, TBH, I don't think it matters that much, leave it ;)